book/src/template_expansion.md

```diff
@@ -0,0 +1,484 @@
+        ::askama::Template::render_into(self, f).map_err(|_| ::std::fmt::Error {})
+    }
+}
+```
+
+For simplicity, we will now only keep the content of the `askama::Template::render_into`
+function from now on.
+
+## Text content
+
+If you have "text content" (for example HTML) in your template:
+
+```html
+<h1>{{ title }}</h1>
+```
+
+It will generate it like this:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "<h1>{0}</h1>",
+            &::askama::MarkupDisplay::new_unsafe(&(self.title), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+About `MarkupDisplay`: we need to use this trait in order to prevent generating
+invalid HTML. Let's take an example: if `title` is `"<a>"` and we display it as
+is, in the generated HTML, you won't see `<a>` but instead a new HTML element
+will be created. To prevent this, we need to escape some characters.
+
+In this example, `<a>` will become `&lt;a&gt;`. And this is why there is the
+`safe` builtin filter, in case you want it to be displayed as is.
+
+## Variables
+
+### Variables creation
+
+If you create a variable in your template, it will be created in the generated
+Rust code as well. For example:
+
+```jinja
+{% set x = 12 %}
+{% let y = x + 1 %}
+```
+
+will generate:
+
+```rust
+let x = 12;
+let y = x + 1;
+::askama::Result::Ok(())
+```
+
+### Variables usage
+
+By default, variables will be fields from the type on which the `askama::Template`
+trait is implemented:
+
+```jinja
+{{ y }}
+```
+
+will generate:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "{0}",
+            &::askama::MarkupDisplay::new_unsafe(&(self.y), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+This is why if the variable is undefined, it won't work with askama and why
+we can't check if a variable is defined or not.
+
+However, you can still access constants and statics by using paths. Let's say
+you have in your Rust code:
+
+```rust
+const FOO: u32 = 0;
+```
+
+Then you can use them in your template by referring to them with a path:
+
+```jinja
+{{ crate::FOO }}{{ super::FOO }}{{ self::FOO }}
+```
+
+It will generate:
+
+```rust
+writer
+    .write_fmt(
+        format_args!(
+            "{0}{1}{2}",
+            &::askama::MarkupDisplay::new_unsafe(&(crate::FOO), ::askama::Html),
+            &::askama::MarkupDisplay::new_unsafe(&(super::FOO), ::askama::Html),
+            &::askama::MarkupDisplay::new_unsafe(&(self::FOO), ::askama::Html),
+        ),
+    )?;
+::askama::Result::Ok(())
+```
+
+Quite reminder: `crate::` is to get an item at the root level of the crate,
+`super::` is to get an item in the parent module and `self::` is to get an item
+in the current module.
+
+You can also access items from your current type as well using `Self::`, it'll
+use the same logic.
+
+## Control blocks
+
+### if/else
+
+There is an interesting twist for `if/else` conditions. Let's check this code:
```

Comment:
"interesting twist" is a little vague. I'd either skip that and be more generic to say "Let's look at code generation for `if`/`else` conditions:" or say something like "The generated code can be more complex than expected, as seen with `if`/`else` conditions:".